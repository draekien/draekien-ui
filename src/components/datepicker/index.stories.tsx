import React from 'react';
import { Story, Meta } from '@storybook/react';
import DatePicker, { DatePickerProps } from '.';

export default {
  title: 'Components/DatePicker',
  component: DatePicker,
} as Meta;

const Template: Story<DatePickerProps> = (args) => <DatePicker {...args} />;

export const Basic = Template.bind({});
Basic.args = {};
