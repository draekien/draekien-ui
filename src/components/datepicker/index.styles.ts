import { ThemeUIStyleObject } from 'theme-ui';

export const datePickerCss: ThemeUIStyleObject = {
  display: 'inline-block',
  padding: '1.5rem',
  width: '17rem',
};

export const datePickerHeaderCss: ThemeUIStyleObject = {
  alignItems: 'center',
  display: 'flex',
  flexDirection: 'row',
  paddingBottom: '1rem',
  justifyContent: 'space-between',
};

export const datePickerLabelCss: ThemeUIStyleObject = {
  color: 'p-300',
  display: 'inline-block',
  textAlign: 'center',
};

export const datePickerWeekdayWrapperCss: ThemeUIStyleObject = {
  color: 'b-400',
  display: 'flex',
  justifyContent: 'space-around',
  marginBottom: '10px',
  textAlign: 'center',
};

export const datePickerWeekdayCss: ThemeUIStyleObject = {
  display: 'inline-block',
  width: '2rem',
};

export const datePickerDayWrapperCss: ThemeUIStyleObject = {
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'space-around',
};

type DatePickerDayCssProps = {
  isSelected: boolean;
  isWithinSelectedRange: boolean;
  isWithinHoverRange: boolean;
  isDisabled: boolean;
  isAnotherMonth?: boolean;
};

export const datePickerDayCss = (
  props: DatePickerDayCssProps
): ThemeUIStyleObject => {
  const css: any = {
    border: '0',
    display: 'inline-block',
    height: '2rem',
    padding: '8px',
    textAlign: 'center',
    variant: 'text.small',
    width: '2rem',
    transition: 'all 300ms',
  };

  if (!props.isDisabled) {
    css.cursor = 'pointer';
    css[':hover'] = {
      backgroundColor: 'p-200',
      color: 'white',
    };
  }

  if (props.isSelected) {
    css.backgroundColor = 'p-300';
    css.color = 'white';
  } else if (props.isWithinSelectedRange || props.isWithinHoverRange) {
    css.backgroundColor = 'p-100';
    css.color = 'b-600';
  } else if (props.isDisabled) {
    css.backgroundColor = 'b-200';
    css.color = 'b-300';
  } else {
    css.backgroundColor = 'p-000';
    css.color = props.isAnotherMonth ? 'b-300' : 'b-600';
  }

  return css;
};
