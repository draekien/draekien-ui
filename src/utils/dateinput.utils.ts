/** THe mask to be used for DateInput */
export const inputMask = 'DD / MM / YYYY';

/** the character used in the mask when the user has not filled the input in */
export const maskChar = '-';

/**
 * Gets a string in the format 'DD / MM / YYYY' from a date
 * @param date
 */
export const getMaskedDateString = (date?: Date) => {
  if (!date || typeof date.getTime !== 'function') return '';

  const month = ('0' + (date.getMonth() + 1)).slice(-2);
  const day = ('0' + date.getDate()).slice(-2);

  return `${day} / ${month} / ${date.getFullYear()}`;
};

/** Gets the string value representing the text the use has entered without the mask characters */
export const getCleanString = (value: string) => {
  if (!value || typeof value !== 'string') return '';
  return value.replace(/\s|-/g, '').replace(/^\/|\/$|\/\//g, '');
};

/** Tries to create a Date object based on a string in the format `DD / MM / YYYY`
 * If is not possible to parse the string value as a date, it will return null
 */
export const getDateFromString = (value?: string) => {
  if (!value || typeof value !== 'string') return null;
  const dateValues = getCleanString(value)
    .split('/')
    .filter((v) => v !== '' && v.length >= 'MM'.length);
  if (dateValues.length < 3) return null;
  if (dateValues[2].length < 4) return null;
  try {
    const numberValues = dateValues.map((v) => parseInt(v)); // [0] day, [1] month, [2] year
    return new Date(numberValues[2], numberValues[1] - 1, numberValues[0]);
  } catch (error) {
    return null;
  }
};

/** Fires the change event for an input element programmatically
 * https://stackoverflow.com/questions/42550341/react-trigger-onchange-if-input-value-is-changing-by-state/42554283
 */
export const fireInputChange = (element: HTMLInputElement, value: string) => {
  const valueSetter = Object.getOwnPropertyDescriptor(element, 'value')?.set;
  const prototype = Object.getPrototypeOf(element);
  const prototypeValueSetter = Object.getOwnPropertyDescriptor(
    prototype,
    'value'
  )?.set;

  if (
    valueSetter &&
    prototypeValueSetter &&
    valueSetter !== prototypeValueSetter
  )
    prototypeValueSetter.call(element, value);
  else if (valueSetter) valueSetter.call(element, value);

  element.dispatchEvent(new Event('input', { bubbles: true }));
};

export const isValidDateInstance = (date?: Date) => {
  return date instanceof Date && !isNaN(date.getTime());
};
