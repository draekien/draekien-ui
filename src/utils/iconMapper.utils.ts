import { colors } from '../components/theme/colors';

export type ValidationColor = 'information' | 'success' | 'warning' | 'error';

export const iconMap: { [key in ValidationColor]: string } = {
  information: 'info',
  success: 'check_circle',
  warning: 'warning',
  error: 'error',
};

export const iconColorMap: { [key in ValidationColor]: keyof typeof colors } = {
  information: 'information-400',
  success: 'success-400',
  warning: 'warning-400',
  error: 'error-400',
};
