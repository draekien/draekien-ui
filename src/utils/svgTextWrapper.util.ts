export const getTextWidth = (text: string, font = '400 1rem sans-serif') => {
  const canvas = document.createElement('canvas');
  const context = canvas.getContext('2d');
  if (!context) return canvas.width;
  context.font = font;
  return context.measureText(text).width;
};

export const breakString = (
  word: string,
  maxWidth: number,
  hyphenCharacter = '-'
) => {
  const characters = word.split('');
  const lines: string[] = [];
  let currentLine = '';

  characters.forEach((character, index) => {
    const nextLine = `${currentLine}${character}`;
    const lineWidth = getTextWidth(nextLine);

    if (lineWidth >= maxWidth) {
      const currentCharacter = index + 1;
      const isLastLine = characters.length === currentCharacter;
      const hyphenatedNextLine = `${nextLine}${hyphenCharacter}`;
      lines.push(isLastLine ? nextLine : hyphenatedNextLine);
      currentLine = '';
    } else {
      currentLine = nextLine;
    }
  });

  return { hyphenatedStrings: lines, remainingWord: currentLine };
};

export const wrapLabel = (
  label: string,
  maxWidth: number,
  truncate = false
) => {
  const words = label.split(' ');
  const completedLines: string[] = [];
  let nextLine = '';

  words.forEach((word, index) => {
    const wordLength = getTextWidth(`${word} `);
    const nextLineLength = getTextWidth(nextLine);

    if (wordLength > maxWidth) {
      if (truncate) {
        const { hyphenatedStrings } = breakString(word, maxWidth, '...');
        completedLines.push(nextLine, hyphenatedStrings[0]);
        return completedLines;
      }

      const { hyphenatedStrings, remainingWord } = breakString(word, maxWidth);
      completedLines.push(nextLine, ...hyphenatedStrings);
      nextLine = remainingWord;
    } else if (nextLineLength + wordLength >= maxWidth) {
      completedLines.push(nextLine);
      nextLine = word;
    } else {
      nextLine = [nextLine, word].filter(Boolean).join(' ');
    }

    const currentWord = index + 1;
    const isLastWord = currentWord === words.length;

    if (isLastWord) {
      completedLines.push(nextLine);
    }
  });

  return completedLines.filter((line) => line !== '');
};
