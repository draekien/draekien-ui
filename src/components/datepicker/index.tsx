/** @jsx jsx */
import { Box, jsx } from 'theme-ui';
import * as React from 'react';
import * as styles from './index.styles';
import * as utils from '../../utils/datepicker.utils';
import Button from '../button';
import Text from '../text';

export type WeekStartDay = 'Sunday' | 'Monday';

export type DatePickerProps = {
  /**
   * An array of dates to pre-select on the datepicker.
   * You must manually validate these dates to be within minDate, maxDate, and maxSelections constraints
   *
   * If mode is range, the first two dates are selected (and those between are highlighted)
   * If mode is multi, all dates are selected
   * If mode is single, only the first date is selected
   *
   * All dates will be stripped of their time
   * @default []
   */
  selectedDates?: Date[];
  /** Minimum selectable date. All dates before this date are disabled */
  minDate?: Date;
  /** Maximum selectable date. All dates after this date are disabled */
  maxDate?: Date;
  /** Initial month when the calendar is rendered for the first time */
  initialMonth?: Date;
  /**
   * Callback handler fro when a date is selected or unselected
   * @param selectedDates array of all the selected dates.
   * In range mode, the first value will be the start date and the second value the end date
   * @param selectedDate the day the user has selected or unselected
   */
  onChange?: (selectedDates: Date[], selectedDate: Date) => void;
  /**
   * Callback handler for when the user moves between months
   * @param activeMonth the active month on the calendar
   */
  onChangeMonth?: (activeMonth: Date) => void;
  /**
   * The mode to run the date picker in
   * range: a start and end date can be selected, with all dates between highlighted
   * multi: a number of dates can be selected up to and including the maxSelections number
   * single: only one date can be selected
   */
  mode?: 'range' | 'multi' | 'single';
  /**
   * Maximum number of selectable dates in multi mode. All unselected dates are disabled when maxSelections is reached
   * @default 10
   */
  maxSelections?: number;
  /**
   * What day the week starts on (Sunday or Monday)
   * @default 'Sunday'
   */
  weekStartDay?: WeekStartDay;
};

export const DatePicker: React.FC<DatePickerProps> = ({
  minDate,
  maxDate,
  initialMonth,
  onChange,
  onChangeMonth,
  mode = 'single',
  maxSelections = 10,
  weekStartDay = 'Sunday',
  ...rest
}) => {
  const [selectedDates, setSelectedDates] = React.useState<Date[]>(
    (rest.selectedDates &&
      utils.clearHoursArray(utils.sortDates(rest.selectedDates))) ||
      []
  );
  const [hoveredDate, setHoveredDate] = React.useState<Date | undefined>(
    undefined
  );
  const [activeMonth, setActiveMonth] = React.useState<Date>(
    initialMonth ||
      (selectedDates.length && utils.getLatestDate(selectedDates)) ||
      new Date()
  );
  const [focusedDate, setFocusedDate] = React.useState<
    'start_date' | 'end_date'
  >('start_date');

  React.useEffect(() => {
    setSelectedDates(
      (rest.selectedDates &&
        utils.clearHoursArray(utils.sortDates(rest.selectedDates))) ||
        []
    );
    setActiveMonth(
      initialMonth ||
        (rest.selectedDates &&
          rest.selectedDates.length &&
          utils.getLatestDate(rest.selectedDates)) ||
        new Date()
    );

    if (
      mode === 'range' &&
      rest.selectedDates &&
      rest.selectedDates.length > 0
    ) {
      setFocusedDate(
        rest.selectedDates.length === 1 ? 'end_date' : 'start_date'
      );
    }
  }, [utils.getDateStrings(rest.selectedDates)]);

  const handleSelect = (date: Date) => {
    let dates;

    switch (mode) {
      case 'single':
        dates = selectedDates[0] === date ? [] : [date];
        break;
      case 'multi':
        dates = utils.addOrRemoveSelectedDate(date, selectedDates);
        break;
      case 'range':
        dates = selectedDates;

        if (focusedDate === 'start_date') {
          dates = [date];
          setFocusedDate('end_date');
        } else {
          dates = utils.sortDates([...dates, date]);
          setFocusedDate('start_date');
        }
        break;
    }

    setSelectedDates(dates);
    onChange && onChange(dates, date);
  };

  const handleOnChangeMonth = (activeMonth: Date) => {
    onChangeMonth && onChangeMonth(activeMonth);
  };

  const handleOnPrevMonth = () => {
    setActiveMonth(utils.subtractMonth(activeMonth));
    handleOnChangeMonth(utils.subtractMonth(activeMonth));
  };

  const handleOnNextMonth = () => {
    setActiveMonth(utils.addMonth(activeMonth));
    handleOnChangeMonth(utils.addMonth(activeMonth));
  };

  const Day = ({ date }: { date: Date }) => {
    const isSelected = utils.isSelected(date, selectedDates);

    const isWithinSelectedRange =
      mode === 'range' && utils.isWithinSelectedRange(date, selectedDates);

    const isWithinHoverRange =
      focusedDate === 'end_date' &&
      utils.isWithinHoverRange(date, selectedDates, hoveredDate);

    const isDisabled =
      (mode === 'multi' &&
        selectedDates.length === maxSelections &&
        !isSelected) ||
      utils.isDisabled(date, minDate, maxDate);

    const isAnotherMonth = date.getMonth() !== activeMonth.getMonth();

    const handleOnClickDay = () => {
      !isDisabled && handleSelect(date);

      if (isAnotherMonth && !isDisabled) {
        setActiveMonth(date);
        handleOnChangeMonth(date);
      }
    };

    return (
      <Box
        as="span"
        onClick={handleOnClickDay}
        onMouseEnter={() => {
          if (
            !isDisabled &&
            (!hoveredDate || date.toISOString() !== hoveredDate.toISOString())
          ) {
            setHoveredDate(date);
          }
        }}
        sx={styles.datePickerDayCss({
          isSelected,
          isWithinSelectedRange,
          isWithinHoverRange,
          isDisabled,
          isAnotherMonth,
        })}
      >
        {date.getDate()}
      </Box>
    );
  };

  return (
    <Box sx={styles.datePickerCss}>
      <Box sx={styles.datePickerHeaderCss}>
        <Button
          onClick={handleOnPrevMonth}
          icon="chevron_left"
          iconPosition="left"
          variant="text"
        />
        <Text as="span" sx={styles.datePickerLabelCss} variant="subtitle">
          {utils.getMonthLabel(activeMonth)}
        </Text>
        <Button
          onClick={handleOnNextMonth}
          icon="chevron_right"
          iconPosition="right"
          variant="text"
        />
      </Box>
      <Box sx={styles.datePickerWeekdayWrapperCss}>
        {utils.getDaysOfWeek(weekStartDay).map((day) => (
          <Text
            as="span"
            key={day.longLabel}
            sx={styles.datePickerWeekdayCss}
            variant="surtitle"
          >
            {day.shortLabel}
          </Text>
        ))}
      </Box>
      <Box sx={styles.datePickerDayWrapperCss}>
        {utils.getDaysOfMonth(activeMonth, weekStartDay).map((day, index) => (
          <Day date={day} key={`${day.toString()}-${index}`} />
        ))}
      </Box>
    </Box>
  );
};

export default DatePicker;
