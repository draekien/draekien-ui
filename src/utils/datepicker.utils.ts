import { WeekStartDay } from '../components/datepicker';

/**
 * Get the days of the week
 * @param weekStartDay which day to start the week with
 */
export const getDaysOfWeek = (weekStartDay: WeekStartDay) => {
  let daysOfWeek = [
    { longLabel: 'Monday', shortLabel: 'M' },
    { longLabel: 'Tuesday', shortLabel: 'T' },
    { longLabel: 'Wednesday', shortLabel: 'W' },
    { longLabel: 'Thursday', shortLabel: 'T' },
    { longLabel: 'Friday', shortLabel: 'F' },
    { longLabel: 'Saturday', shortLabel: 'S' },
  ];
  const sunday = { longLabel: 'Sunday', shortLabel: 'S' };
  if (weekStartDay === 'Sunday') daysOfWeek = [sunday, ...daysOfWeek];
  else daysOfWeek.push(sunday);
  return daysOfWeek;
};

/**
 * Array of the months of the year
 */
export const MONTHS_OF_YEAR = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

/**
 * Sorts an array of dates from smallest to largest
 * @param dates array of dates to sort
 */
export const sortDates = (dates: Date[]) => {
  const newDates = [...dates];
  return newDates.sort((a, b) => a.getTime() - b.getTime());
};

/**
 * Sorts an array of dates from smallest to largest and returns the largest date
 * @param dates array of dates to sort
 */
export const getLatestDate = (dates: Date[]) => {
  const newDates = [...dates];
  return sortDates(newDates)[newDates.length - 1];
};

/**
 * Adds a date to an array if it does not exist, or removes it if it does
 * @param date date to add or remove
 * @param dates existing array of dates
 * @returns the new array sorted from smallest to largest date
 */
export const addOrRemoveSelectedDate = (date: Date, dates: Date[]) => {
  const newDates = [...dates];
  const index = newDates.findIndex((newDate) => +newDate === +date);
  if (index > -1) newDates.splice(index, 1);
  else newDates.push(date);
  return sortDates(newDates);
};

/**
 * Sets the hours of an array of dates to midnight
 * @param dates array of dates
 */
export const clearHoursArray = (dates: Date[]) => {
  const newDates: Date[] = [];
  dates.forEach((date) => newDates.push(clearHours(date)));
  return newDates;
};

/**
 * Sets the hours of a date to midnight
 * @param date
 */
export const clearHours = (date: Date) => {
  const newDate = new Date(date);
  newDate.setHours(0, 0, 0, 0);
  return newDate;
};

/**
 * Sets the month of a date to the previous month
 * @param date
 */
export const subtractMonth = (date: Date) => {
  const newDate = new Date(date);
  newDate.setMonth(date.getMonth() - 1);
  return newDate;
};

/**
 * Sets the month of a date to the next month
 * @param date
 */
export const addMonth = (date: Date) => {
  const newDate = new Date(date);
  newDate.setMonth(date.getMonth() + 1);
  return newDate;
};

/**
 * Gets a month-year label for a provided date
 * @param date
 */
export const getMonthLabel = (date: Date) => {
  return `${MONTHS_OF_YEAR[date.getMonth()]} ${date.getFullYear()}`;
};

/**
 * Gets the first date of the month for a given date
 * @param date
 */
export const getFirstDateOfMonth = (date: Date) => {
  return new Date(date.getFullYear(), date.getMonth(), 1);
};

/**
 * Gets the last date of the month for a given date
 * @param date
 */
export const getLastDateOfMonth = (date: Date) => {
  return new Date(date.getFullYear(), date.getMonth() + 1, 0);
};

/**
 * Gets the first day of the month for a given date
 * @param date
 * @param weekStartDay what day the week starts on
 */
export const getFirstDayOfMonth = (date: Date, weekStartDay: WeekStartDay) => {
  let firstDayOfMonth = getFirstDateOfMonth(date).getDay();
  if (weekStartDay === 'Monday') firstDayOfMonth -= 1;
  if (firstDayOfMonth === -1) firstDayOfMonth = 6;
  return firstDayOfMonth;
};

/**
 * Gets the last day of the month for a given date
 * @param date
 * @param weekStartDay what day the week starts on
 */
export const getLastDayOfMonth = (date: Date, weekStartDay: WeekStartDay) => {
  let lastDayOfMonth = getLastDateOfMonth(date).getDay();
  if (weekStartDay === 'Monday') lastDayOfMonth -= 1;
  if (lastDayOfMonth === -1) lastDayOfMonth = 6;
  return lastDayOfMonth;
};

/**
 * Gets the dates to display in a calendar month, including previous and next months to fill in full weeks
 * @param date
 * @param weekStartDay what day the week starts on
 */
export const getDaysOfMonth = (date: Date, weekStartDay: WeekStartDay) => {
  const days = [];

  // Fill in the top row with the last few days of last month if required
  for (let i = getFirstDayOfMonth(date, weekStartDay); i > 0; i--) {
    const newDate = getFirstDateOfMonth(date);
    newDate.setDate(newDate.getDate() - i);
    days.push(clearHours(newDate));
  }

  // Fill in the month
  for (
    let i = getFirstDateOfMonth(date);
    i <= getLastDateOfMonth(date);
    i.setDate(i.getDate() + 1)
  ) {
    days.push(clearHours(new Date(i)));
  }

  // Fill in the bottom row with the first few days of next month if required
  for (let i = 1; i <= 6 - getLastDayOfMonth(date, weekStartDay); i++) {
    const newDate = getLastDateOfMonth(date);
    newDate.setDate(newDate.getDate() + i);
    days.push(clearHours(newDate));
  }

  return days;
};

/**
 * Checks if a date is a selected date
 * @param date
 * @param selectedDates
 */
export const isSelected = (date: Date, selectedDates: Date[]) =>
  selectedDates.some(
    (selectedDate) => selectedDate.toISOString() === date.toISOString()
  ) || false;

/**
 * Checks if a date falls within the range of two selected dates
 * @param date
 * @param selectedDates
 */
export const isWithinSelectedRange = (date: Date, selectedDates: Date[]) =>
  (selectedDates[0] &&
    date >= selectedDates[0] &&
    selectedDates[1] &&
    date <= selectedDates[1]) ||
  false;

/**
 * Checks if a date falls within the range of a selected date and a hovered date
 * @param date
 * @param selectedDates
 * @param hoveredDate
 */
export const isWithinHoverRange = (
  date: Date,
  selectedDates: Date[],
  hoveredDate?: Date
) => {
  if (selectedDates[0] && hoveredDate) {
    if (hoveredDate > selectedDates[0]) {
      return date >= selectedDates[0] && date <= hoveredDate;
    } else {
      return date >= hoveredDate && date <= selectedDates[0];
    }
  }
  return false;
};

/**
 * Checks if a date falls outside the min and max date range
 * @param date
 * @param minDate
 * @param maxDate
 */
export const isDisabled = (date: Date, minDate?: Date, maxDate?: Date) =>
  (minDate && date < clearHours(minDate)) ||
  (maxDate && date > clearHours(maxDate)) ||
  false;

export const getDateStrings = (dates?: Date[]) =>
  dates
    ? dates
        .map((d) => [d.getDate(), d.getMonth(), d.getFullYear()].join())
        .join()
    : '';
