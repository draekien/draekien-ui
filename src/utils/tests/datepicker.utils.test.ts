import * as utils from '../datepicker.utils';

const dates = [
  new Date(2020, 5, 1),
  new Date(2020, 1, 1),
  new Date(2020, 2, 1),
  new Date(2020, 12, 1),
];

describe('DatePicker util', () => {
  test('getDaysOfWeek should return days of week in correct order when start day is Monday', () => {
    const daysOfWeek = utils.getDaysOfWeek('Monday');

    expect(daysOfWeek[0]).toEqual({ longLabel: 'Monday', shortLabel: 'M' });
    expect(daysOfWeek).toMatchSnapshot();
  });

  test('getDaysOfWeek should return days of week in correct order when start day is Sunday', () => {
    const daysOfWeek = utils.getDaysOfWeek('Sunday');

    expect(daysOfWeek[0]).toEqual({ longLabel: 'Sunday', shortLabel: 'S' });
    expect(daysOfWeek).toMatchSnapshot();
  });

  test('sortDates should sort dates from smallest to largest', () => {
    const sortedDates = utils.sortDates(dates);

    expect(sortedDates[0]).toEqual(new Date(2020, 1, 1));
    expect(sortedDates[1]).toEqual(new Date(2020, 2, 1));
    expect(sortedDates[2]).toEqual(new Date(2020, 5, 1));
    expect(sortedDates[3]).toEqual(new Date(2020, 12, 1));
  });

  test('getLatestDate should return the largest date', () => {
    const largestDate = utils.getLatestDate(dates);

    expect(largestDate).toEqual(new Date(2020, 12, 1));
  });

  test('addOrRemoveSelectedDate should add date when date does not exist', () => {
    const newDates = utils.addOrRemoveSelectedDate(new Date(2020, 1, 2), dates);

    expect(newDates).toContainEqual(new Date(2020, 1, 2));
  });

  test('addOrRemoveSelectedDate should remove date when date exists', () => {
    const newDates = utils.addOrRemoveSelectedDate(new Date(2020, 1, 1), dates);

    expect(newDates).not.toContainEqual(new Date(2020, 1, 1));
  });

  test('clearHours should set hours to midnight', () => {
    const date = utils.clearHours(new Date(2020, 1, 1, 20));

    expect(date).toEqual(new Date(2020, 1, 1));
  });

  test('clearHoursArray should set all dates to midnight', () => {
    const clearedDates = utils.clearHoursArray([
      new Date(2020, 1, 1, 20),
      new Date(2020, 2, 1, 20),
      new Date(2020, 3, 1, 20),
    ]);

    clearedDates.forEach((date) => {
      expect(date.getHours()).toEqual(0);
    });
  });

  test('substractMonth should substract month', () => {
    const expectedDate = new Date(2020, 3, 1);
    const date = utils.subtractMonth(new Date(2020, 4, 1));

    expect(date).toEqual(expectedDate);
  });

  test('addMonth should add month', () => {
    const expectedDate = new Date(2020, 3, 1);
    const date = utils.addMonth(new Date(2020, 2, 1));

    expect(date).toEqual(expectedDate);
  });

  test('getMonthLabel should return friendly month and year', () => {
    const expectedLabel = 'February 2020';
    const label = utils.getMonthLabel(new Date(2020, 1, 1));

    expect(label).toEqual(expectedLabel);
  });

  test('getFirstDateOfMonth should return first date of month', () => {
    const expectedDate = new Date(2020, 1, 1);
    const date = utils.getFirstDateOfMonth(new Date(2020, 1, 24));

    expect(date).toEqual(expectedDate);
  });

  test('getLastDateOfMonth should return last date of month', () => {
    const expectedDate = new Date(2020, 0, 31);
    const date = utils.getLastDateOfMonth(new Date(2020, 0, 15));

    expect(date).toEqual(expectedDate);
  });

  test('getFirstDayOfMonth should return first day of month when week start day is Sunday', () => {
    const expectedDay = new Date(2020, 0, 1).getDay();
    const day = utils.getFirstDayOfMonth(new Date(2020, 0, 14), 'Sunday');

    expect(day).toEqual(expectedDay);
  });

  test('getFirstDayOfMonth should return first day of month when week start day is Monday', () => {
    const expectedDay = new Date(2020, 0, 1).getDay() - 1;
    const day = utils.getFirstDayOfMonth(new Date(2020, 0, 14), 'Monday');

    expect(day).toEqual(expectedDay);
  });

  test('getLastDayOfMonth should return last day of month when week start day is Sunday', () => {
    const expectedDay = new Date(2020, 0, 31).getDay();
    const day = utils.getLastDayOfMonth(new Date(2020, 0, 14), 'Sunday');

    expect(day).toEqual(expectedDay);
  });

  test('getLastDayOfMonth should return last day of month when week start day is Monday', () => {
    const expectedDay = new Date(2020, 0, 31).getDay() - 1;
    const day = utils.getLastDayOfMonth(new Date(2020, 0, 14), 'Monday');

    expect(day).toEqual(expectedDay);
  });

  test('getDaysOfMonth should match snapshot when week start day is Monday', () => {
    const days = utils.getDaysOfMonth(new Date(2020, 0, 12), 'Monday');

    expect(days).toMatchSnapshot();
  });

  test('getDaysOfMonth should match snapshot when week start day is Sunday', () => {
    const days = utils.getDaysOfMonth(new Date(2020, 0, 12), 'Sunday');

    expect(days).toMatchSnapshot();
  });

  test('isSelected should return true if date is within selected dates', () => {
    const isSelected = utils.isSelected(new Date(2020, 1, 1), dates);

    expect(isSelected).toBeTruthy();
  });

  test('isSelected should return false if date is not in selected dates', () => {
    const isSelected = utils.isSelected(new Date(2020, 3, 1), dates);

    expect(isSelected).toBeFalsy();
  });

  test('isWithinSelectedRange should return true when date is within selected range', () => {
    const range = [new Date(2020, 0, 1), new Date(2020, 1, 1)];
    const date = new Date(2020, 0, 15);
    const isWithinSelectedRange = utils.isWithinSelectedRange(date, range);

    expect(isWithinSelectedRange).toBeTruthy();
  });

  test('isWithinSelectedRange should return false when date is outside selected range', () => {
    const range = [new Date(2020, 0, 1), new Date(2020, 1, 1)];
    const date = new Date(2020, 1, 15);
    const isWithinSelectedRange = utils.isWithinSelectedRange(date, range);

    expect(isWithinSelectedRange).toBeFalsy();
  });

  test('isWithinHoverRange should return true when date is within selected date and hovered date', () => {
    const range = [new Date(2020, 0, 1)];
    const date = new Date(2020, 0, 15);
    const hovered = new Date(2020, 1, 1);

    const isWithinHoverRange = utils.isWithinHoverRange(date, range, hovered);

    expect(isWithinHoverRange).toBeTruthy();
  });

  test('isWithinHoverRange should return false when date is outside selected date and hovered date', () => {
    const range = [new Date(2020, 0, 1)];
    const date = new Date(2020, 1, 15);
    const hovered = new Date(2020, 1, 1);

    const isWithinHoverRange = utils.isWithinHoverRange(date, range, hovered);

    expect(isWithinHoverRange).toBeFalsy();
  });

  test('isWithinHoverRange should return false when hovered date is undefined', () => {
    const range = [new Date(2020, 0, 1)];
    const date = new Date(2020, 1, 15);

    const isWithinHoverRange = utils.isWithinHoverRange(date, range);

    expect(isWithinHoverRange).toBeFalsy();
  });

  test('isDisabled should return correct result when date falls outside minmax date range', () => {
    const minDate = new Date(2020, 1, 1);
    const maxDate = new Date(2020, 1, 10);

    expect(
      utils.isDisabled(new Date(2020, 0, 1, 10), minDate, maxDate)
    ).toBeTruthy();
    expect(
      utils.isDisabled(new Date(2020, 1, 15, 15), minDate, maxDate)
    ).toBeTruthy();
    expect(utils.isDisabled(new Date(2020, 0, 1, 10), minDate)).toBeTruthy();
    expect(utils.isDisabled(new Date(2020, 1, 15, 15), minDate)).toBeFalsy();
    expect(
      utils.isDisabled(new Date(2020, 0, 1, 10), undefined, maxDate)
    ).toBeFalsy();
    expect(
      utils.isDisabled(new Date(2020, 1, 15, 15), undefined, maxDate)
    ).toBeTruthy();
  });

  test('isDisabled should return false when date falls inside minmax date range', () => {
    const minDate = new Date(2020, 1, 1);
    const maxDate = new Date(2020, 1, 10);

    expect(
      utils.isDisabled(new Date(2020, 1, 2, 10), minDate, maxDate)
    ).toBeFalsy();
  });

  test('isDisabled should return false when minmax date is undefined', () => {
    expect(utils.isDisabled(new Date(2020, 1, 2, 10))).toBeFalsy();
  });

  test('getDateStrings returns date string', () => {
    expect(utils.getDateStrings()).toEqual('');
    expect(utils.getDateStrings(dates)).toMatchSnapshot();
  });
});
