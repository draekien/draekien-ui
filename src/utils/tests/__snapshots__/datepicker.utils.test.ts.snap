// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker util getDateStrings returns date string 1`] = `"1,5,2020,1,1,2020,1,2,2020,1,0,2021"`;

exports[`DatePicker util getDaysOfWeek should return days of week in correct order when start day is Monday 1`] = `
Array [
  Object {
    "longLabel": "Monday",
    "shortLabel": "M",
  },
  Object {
    "longLabel": "Tuesday",
    "shortLabel": "T",
  },
  Object {
    "longLabel": "Wednesday",
    "shortLabel": "W",
  },
  Object {
    "longLabel": "Thursday",
    "shortLabel": "T",
  },
  Object {
    "longLabel": "Friday",
    "shortLabel": "F",
  },
  Object {
    "longLabel": "Saturday",
    "shortLabel": "S",
  },
  Object {
    "longLabel": "Sunday",
    "shortLabel": "S",
  },
]
`;

exports[`DatePicker util getDaysOfWeek should return days of week in correct order when start day is Sunday 1`] = `
Array [
  Object {
    "longLabel": "Sunday",
    "shortLabel": "S",
  },
  Object {
    "longLabel": "Monday",
    "shortLabel": "M",
  },
  Object {
    "longLabel": "Tuesday",
    "shortLabel": "T",
  },
  Object {
    "longLabel": "Wednesday",
    "shortLabel": "W",
  },
  Object {
    "longLabel": "Thursday",
    "shortLabel": "T",
  },
  Object {
    "longLabel": "Friday",
    "shortLabel": "F",
  },
  Object {
    "longLabel": "Saturday",
    "shortLabel": "S",
  },
]
`;
